hw/xbox/xid-sdl.c

```diff
@@ -464,6 +446,81 @@ static void usb_xid_class_initfn(ObjectClass *klass, void *data)
     uc->handle_attach  = usb_desc_attach;
 }
 
+static int apply_haptic_effect(USBXIDState *s, SDL_HapticEffect *effect) {
+    int effect_id = SDL_HapticNewEffect(s->sdl_haptic, effect);
+    if (effect_id == -1) {
+        const char* error = SDL_GetError();
+        fprintf(stderr, "Failed to create haptic effect! SDL Error: %s\n", error);
+        SDL_HapticClose(s->sdl_haptic);
+        s->sdl_haptic = NULL;
+        s->sdl_haptic_effect = NULL;
+        return -1;
+    }
+
+    int run_result = SDL_HapticRunEffect(s->sdl_haptic, effect_id, 1);
+    if (run_result == -1) {
+        const char* error = SDL_GetError();
+        fprintf(stderr, "Failed to run haptic effect! SDL Error: %s\n", error);
+        SDL_HapticClose(s->sdl_haptic);
+        s->sdl_haptic = NULL;
+        s->sdl_haptic_effect = NULL;
+        return -1;
+    }
+
+    s->sdl_haptic_effect = effect;
+    s->sdl_haptic_effect_id = effect_id;
+    return 0;
+}
+
+static int haptic_effect_init(USBXIDState *s, SDL_HapticEffect *effect) {
+    if (SDL_HapticQuery(s->sdl_haptic) & SDL_HAPTIC_LEFTRIGHT) {
```

Comment:
We should probably have some `EmulationMode` enum (bad name) which picks which method to use for our effects on startup. That way, we only have to do the query once.

We'd then store the chosen emulation mode in `s` to generate the proper effect structure.

This is more extensible and can also be used in the future to take a user-override for devices with broken force feedback.

Reply:
I think with that method we can also avoid allocating on the heap. I like it :+1:

Reply:
Actually SDL also keeps their stuff on the heap.. https://github.com/spurious/SDL-mirror/blob/master/src/haptic/SDL_haptic.c#L821

Just find a working solution which is easy to read (I still prefer the heapless variant)